### PR TITLE
LibWeb: Various style/layout invalidation avoidance

### PR DESCRIPTION
**Pull request by awesomekling**
See individual commits.

Loading https://github.com/LadybirdBrowser/ladybird
Before: 12s
After: 8.5s

### CODE REVIEW

File: Userland/Libraries/LibWeb/DOM/Node.cpp [hunk 1]
```diff
@@ -656,7 +656,12 @@ void Node::insert_before(JS::NonnullGCPtr<Node> node, JS::GCPtr<Node> child, boo
...
         // FIXME: This will need to become smarter when we implement the :has() selector.
         invalidate_style(StyleInvalidationReason::NodeInsertBefore);
-        document().invalidate_layout_tree();
+
+        // NOTE: If the parent didn't have a layout node before, rebuilding the layout tree isn't gonna give it one
+        //       just because it got a new child.
+        if (layout_node()) {
+            document().invalidate_layout_tree();
+        }
     }
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/DOM/Node.cpp]: I don't think it's going to work for elements with "display: contents". for example here insertion of new div.item into div.container should invalidate layout tree even though div.container does not have a layout node:
```html
<!DOCTYPE html>
<style>
  .container {
    display: contents;
  }

  .item {
    padding: 10px;
    margin: 5px;
    border: 1px solid darkblue;
  }
</style>
<div class="container">
  <div class="item">Item 1</div>
  <div class="item">Item 2</div>
  <div class="item">Item 3</div>
</div>

```

Comment (awesomekling) [reply Userland/Libraries/LibWeb/DOM/Node.cpp]: Hmm damn, good point. What about if neither parent nor siblings have layout nodes?

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/DOM/Node.cpp]: it won't word for nested "display: contents" without siblings like:
```html
<!DOCTYPE html>
<style>
  .container {
    display: contents;
  }

  .item {
    padding: 10px;
    margin: 5px;
    border: 1px solid darkblue;
  }
</style>
<div class="container">
<div class="container">
<div class="container">
<div class="container">
  <div class="item">Item 1</div>
  <div class="item">Item 2</div>
  <div class="item">Item 3</div>
</div>
```

maybe simply checking if a node has "display: contents" will work?

Comment (awesomekling) [reply Userland/Libraries/LibWeb/DOM/Node.cpp]: I'll just take this commit out for now, it barely had any impact anyway. We'll come back to it with a clearer idea of what invalidation is really needed :)

**Pull request merged**